### ISSUE TITLE
Feature: Add common DI engine adapters.

### ISSUE THREAD

**Issue by glennawatson** [reactiveui/splat#233]
**Is your feature request related to a problem? Please describe.**
Add new NuGet packages for different DI engines, such as Autofac, SimpleInjector etc.

**Describe the solution you'd like**
Add under a NuGet Package scheme, such as Splat.DependencyInjection.SimpleInjector or something.

**Describe suggestions on how to achieve the feature**
A clear description to how to achieve the feature.
Gives users the chance to use popular DI engines.

- [x] Autofac
- [x] Ninject
- [x] ~~Structure Map~~
- [x] Simple Injector
- [ ] Prism
- [x] DryIoc
- [ ] Microsoft.Extensions.DependencyInjection

Relates to https://github.com/reactiveui/rfcs/issues/14

**Comment by glennawatson** [comment]
This relates to a conversation started in the reactiveui/website project. Instead of documenting the different approaches we probably should just provide adapters.

**Comment by glennawatson** [comment]
once someone does this one https://github.com/reactiveui/splat/issues/233 we can change this documentation dramatically.

**Comment by RLittlesII** [comment]
Structure Map has been [sunset](https://jeremydmiller.com/2018/01/29/sunsetting-structuremap/)

**Comment by chuuddo** [comment]
SimpleInjectorDependencyResolver not works in real apps. 
I made a pull request #239 with changes to View and ViewModel classes to start initialization of RxApp and all tests fails.

After some investigations i found this issues:
1. [The container is locked after the first call to resolve](https://simpleinjector.readthedocs.io/en/latest/decisions.html#the-container-is-locked-after-the-first-call-to-resolve) but ReactiveUI invoke all registrations after RxApp.EnsureInitialized(). This can be fixed by not getting instace of MainWindow from container. If there are no other solutions, we should add this workaround to documentation.
```csharp
    //var window = (MainWindow)container.GetInstance<IViewFor<MainViewModel>>();
    var window = new MainWindow();
    window.ViewModel = container.GetInstance<MainViewModel>();
    window.Show();
```

2. SimpleInjector support multiple registrations for same interface only through [Collection.Register](https://simpleinjector.readthedocs.io/en/latest/using.html#collections). We can't register [this lines](https://github.com/reactiveui/ReactiveUI/blob/ea99102c37bbbf2567c0193e558d1ef9402fdaed/src/ReactiveUI/Registration/Registrations.cs#L23-L25) using _container.Register() method.

https://github.com/reactiveui/splat/blob/d27b0b48e71d38402c8e224b25e2dfff2daef875/src/Splat.SimpleInjector/SimpleInjectorDependencyResolver.cs#L38

I don't have any workarounds for second issue. Can we use this DI container with ReactiveUI?

**Comment by glennawatson** [comment]
@chuuddo Thanks for the comments. Might be worth moving this one into it's own issue. Also if you can maybe provide a small project with the issue it'll like Rodney's life easier.

**Comment by glennawatson** [comment]
@chuuddo Looks like you've done better than a small demo :) But the issue may be better to link to yourPR.

**Comment by chuuddo** [comment]
Just looked at Autofac implementation. 
Autofac container is also [immutable](https://autofaccn.readthedocs.io/en/latest/best-practices/index.html?highlight=update#consider-a-container-as-immutable) after build. 
So, we can't use ReactiveUI with immutable DI containers without resolving https://github.com/reactiveui/ReactiveUI/issues/1737.
Or use mutable DI containers like DryIoc, Ninject.

**Comment by RLittlesII** [comment]
Thanks for the Simple Injector catch.  It may not be compatible with ReactiveUI, I'll look at the findings and see if we can find a reasonable way to resolve it.

Im confused why muttability matters?  As long as you aren't registering elements after the container is built there shouldn't be any issue using Autofac.

There is a link in the Autofac implementation explaining that you would have to explicitly rebuild the container.

This puts the onus on the developer to know that they are deviating from the frameworks in ended use, and that they have to manage it themselves.

Ninject is on the list of implementations, we just haven't gotten to it yet.  I can add DryIoc.

Thanks again for the detail!

**Comment by chuuddo** [comment]
> Im confused why muttability matters? As long as you aren't registering elements after the container is built there shouldn't be any issue using Autofac.

Mutability matters because of current implementation of ReactiveUI initialization.
We build Autofac container before RxApp.EnsureInitialized(), thats why we cant register all RxApp stuff without builder.Update(container), but that method is obsolete.

And it looks like this method missed in current AutofacDependencyResolver implementation.
https://github.com/reactiveui/splat/blob/d27b0b48e71d38402c8e224b25e2dfff2daef875/src/Splat.Autofac/AutofacDependencyResolver.cs#L73-L84

**Comment by RLittlesII** [comment]
Yes I removed it because it is obsolete.  I plan on migrating an app to this implementation this week.  If I have issues, I will update this implementation.

I am explicitly calling the Splat and ReactiveUI initialization before I assign the container to splat.  I haven't had any trouble with it so far.

**Comment by chuuddo** [comment]
Provide another repo with some tests. Hope it will be useful.
https://github.com/chuuddo/SplatDependencyResolversTests

**Comment by worldbeater** [comment]
Adding support for `Microsoft.Extensions.DependencyInjection` as well might be [a good idea](https://github.com/reactiveui/website/issues/292#issuecomment-460561529).

**Comment by glennawatson** [comment]
hey @chuuddo -- the static locator is bad in tests, we have a fix coming in #282 which makes the Locator available as a class for testing. If you want to do a PR with your suggested tests feel free to pull a new PR after #282 has been merged.

**Comment by chuuddo** [comment]
@glennawatson i can make a PR, but in #274 you want to keep RxUI out of Splat and i want to clarify do we need to add this tests at all.

**Comment by glennawatson** [comment]
Any remaining tasks with this one @RLittlesII

**Comment by RLittlesII** [comment]
@glennawatson I believe you implemented the `Microsoft.Extensions.DependencyInjection`.  That just leaves Prism.  I would be fine closing this issue and opening one up specifically for Prism DI, and let the community implement it if they determine they want it.  At this point we have enough examples of how to do it, should be a good first issue for someone.

**Comment by glennawatson** [comment]
Sounds good, mark it as first timer with a label I guess.
